### PR TITLE
fix(whale-api): add reindex for failed to indexed block

### PR DESCRIPTION
**Pull request by lykalabrada**
#### What this PR does / why we need it:

Fixing random bug `!bestChain` 

- introduce status `reindex` to jump off the cleanup cycle if `NotFoundIndexError`
- logged on random bug`!bestChain`

#### Which issue(s) does this PR fixes?:

Fixes #

### CODE REVIEW

**Comment by canonbrother** [comment]
@lykalabrada 
suggestion: 
```diff
-- await this.statusMapper.put(hash, height, Status.ERROR)
++ if (err instanceof NotFoundIndexerError) {
++         await this.statusMapper.put(hash, height, Status.REINDEX)
++ } else {
++         await this.statusMapper.put(hash, height, Status.ERROR)
++ }
```

```diff
// cleanup
...
switch (status.status) {
           case Status.INVALIDATED:
           case Status.INDEXED:
++       case Status.REINDEX:        
        return
    }
```    
it will solve the cleanup lock and reindex based on bestchain checker.. at the end.. we'd finally see the "issue" log

File: apps/whale-api/src/module.indexer/rpc.block.provider.ts [hunk 1]
```diff
@@ -88,9 +98,32 @@ export class RPCBlockProvider {
 
     const nextBlock = await this.client.blockchain.getBlock(nextHash, 2)
     if (await RPCBlockProvider.isBestChain(indexed, nextBlock)) {
+      this.logger.log(`[Synchronize - best chain] Indexing next block ${nextBlock.height}...`)
       await this.index(nextBlock)
     } else {
-      await this.invalidate(indexed.hash, indexed.height)
+      this.logger.log(`[Synchronize - not the best chain] Indexing prev block ${indexed.height}...`)
+      // Retry indexing the previous block before invalidating it
+      const MAX_RETRIES = 3
```
Comment (canonbrother) [inline apps/whale-api/src/module.indexer/rpc.block.provider.ts]: we do not need this ady after `REINDEX`..
all reindex/pre-index will be auto handling in `synchronize`
one `retry pattern` is sufficient

File: apps/whale-api/src/module.indexer/rpc.block.provider.ts [hunk 2]
```diff
@@ -100,7 +133,7 @@ export class RPCBlockProvider {
    * @param {defid.Block<Transaction>} nextBlock to check previous block hash
    */
   private static async isBestChain (indexed: Block, nextBlock: defid.Block<defid.Transaction>): Promise<boolean> {
-    return nextBlock.previousblockhash === indexed.hash
+    return indexed.hash === nextBlock.previousblockhash
```
Comment (canonbrother) [inline apps/whale-api/src/module.indexer/rpc.block.provider.ts:105]: nice 🤓  !! (lesser brain consuming)

File: apps/whale-api/src/module.indexer/rpc.block.provider.ts [hunk 3]
```diff
@@ -44,6 +45,15 @@ export class RPCBlockProvider {
       }
 
       try {
+        const status = await this.statusMapper.get()
```
Comment (canonbrother) [inline apps/whale-api/src/module.indexer/rpc.block.provider.ts]: same as
https://github.com/BirthdayResearch/jellyfishsdk/pull/2169/files#r1388923584

File: apps/whale-api/src/module.indexer/rpc.block.provider.ts (hunk 1 above)
Comment (lykalabrada) [reply apps/whale-api/src/module.indexer/rpc.block.provider.ts]: Updated, then we will not be reindexing the previous block, just the current block thats failing. Also fine with me, we just have to test on any stucked ocean, to see if fix still works the same :)
cc: @pierregee

**Comment by canonbrother** [comment]
i think need a log here tho...
https://github.com/BirthdayResearch/jellyfishsdk/blob/6f312b9b5a6c8c98e0c8418f55914a46b13f90c4/apps/whale-api/src/module.indexer/rpc.block.provider.ts#L92-L93
eg:
```
this.logger.error(`indexed{n: ${indexed.height}, h: ${indexed.hash}}, nextBlock{n: ${nextBlock.height}, prevH: ${nextBlock.previousblockhash}}`)
```

File: apps/whale-api/src/module.indexer/rpc.block.provider.ts [hunk 4]
```diff
@@ -149,7 +152,11 @@ export class RPCBlockProvider {
       await this.indexer.invalidate(hash)
       await this.statusMapper.put(hash, height, Status.INVALIDATED)
     } catch (err) {
-      await this.statusMapper.put(hash, height, Status.ERROR)
+      if (err instanceof NotFoundIndexerError) {
```
Comment (canonbrother) [inline apps/whale-api/src/module.indexer/rpc.block.provider.ts:155]: fix for (composite)poolswap index 
```suggestion
      if (err instanceof NotFoundIndexerError || err instanceof IndexerError) {
```

**Review by canonbrother — APPROVED** [review]

**Pull request merged**